### ISSUE TITLE
Switch to faster HTTP library

### ISSUE THREAD

**Issue by mantas** [zammad/zammad#2880]
### Infos:

* Used Zammad version: 3.2
* Installation method (source, package, ..): any
* Operating system: any
* Database + version: any
* Elasticsearch version: any
* Browser + version: any

### Expected behavior:

* Use faster HTTP library. For example [Typhoeus](https://github.com/typhoeus/typhoeus)

**Comment by thorsteneckel** [comment]
Can you please share some background information on this? What is the issue with using the default `Net::HTTP`?

**Comment by mantas** [comment]
@thorsteneckel cURL-based libraries are quite a bit faster. Especially in multi-threaded environment. I run a couple benchmarks locally and Typhoeus seems to be ~ 2x faster than `Net::HTTP` in Ruby 2.6.5. This could give us a speed boost in search and other places where HTTP connections are run under the hood.

**Comment by thorsteneckel** [comment]
Hey @mantas - I raised this topic while @martini and I discussed the upcoming performance release 3.3. We discussed the advantages and disadvantages of using a third party dependency instead of the build in `Net::HTTP`. We also evaluated the change in the scope of the topic that caused the evaluation (slow search performance).
@martini pointed out that the actual performance issue comes from sending way too much requests to the Elasticsearch when using the global app search in the WebApp (each char * number of searchable models). He has a branch that's fixing this behavior in the pipeline. He also pointed out that the optimization to make parallel requests via threads did not gained a relevant performance improvement (because the number if requests stays the same) and therefore was/will be discarded.
Backed with this gained knowledge I evaluated this proposal. From my point of view this comes with the downsides:

* Research overhead
* Implementation overhead
* Increased number of external dependencies (abandoned, stale, security fixes)
* Maintenance overhead
* Unknown impact on Production and Test environment on various OSs

On the other side we got the `2x` performance improvements. This only affects the actual HTTP requests, right? Can you share some background information about the benchmarks?

TBH I got a feeling that this is a micro optimization for which the downsides predominate. In comparison to the other open issues this comes with a lower customer value.

**Comment by mantas** [comment]
Yes, this only affects requests made using `UserAgent` wrapper. And I agree that it adds some maintenance overhead.

I'd say bumping Ruby version is higher priority. It features improved `Net::HTTP` performance according to the same benchmarks. Unfortunately I didn't save detailed numbers from the benchmarks, but overall cURL-based wrappers are slightly faster than `Net::HTTP` and newer Ruby versions have improved `Net::HTTP` performance, especially when running multi-threaded.

All-in-all I agree this is low priority (bordering `wont-fix`) issue. Especially if there're other ways to solve ES performance issues. But it may be good last ditch effort if we come to that...
